File: src/main/java/kitchenpos/eatinorder/tobe/domain/order/OrderLineItems.java
package kitchenpos.eatinorder.tobe.domain.order;

import jakarta.persistence.*;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

@Embeddable
public class OrderLineItems {

    @OneToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
    @JoinColumn(
            name = "order_id",
            nullable = false,
            columnDefinition = "binary(16)",
            foreignKey = @ForeignKey(name = "fk_order_line_item_to_orders")
    )
    private List<OrderLineItem> orderLineItems = new ArrayList<>();

    protected OrderLineItems() {
    }

    public OrderLineItems(List<OrderLineItem> orderLineItems) {
        validate(orderLineItems);
        this.orderLineItems = new ArrayList<>(orderLineItems);
    }

    private static void validate(List<OrderLineItem> orderLineItems) {
        if (Objects.isNull(orderLineItems) || orderLineItems.isEmpty()) {
            throw new IllegalArgumentException("주문 내역은 빈값일 수 없습니다.");
        }
    }

    public List<OrderLineItem> getOrderLineItems() {
        return Collections.unmodifiableList(orderLineItems);
    }
}

File: src/main/kotlin/kitchenpos/tobe/eatinorder/domain/entity/OrderLineItems.kt
package kitchenpos.tobe.eatinorder.domain.entity

import jakarta.persistence.CascadeType
import jakarta.persistence.Embeddable
import jakarta.persistence.ForeignKey
import jakarta.persistence.JoinColumn
import jakarta.persistence.OneToMany

@Embeddable
class OrderLineItems private constructor(
    @OneToMany(cascade = [CascadeType.PERSIST, CascadeType.MERGE])
    @JoinColumn(
        name = "order_id",
        nullable = false,
        columnDefinition = "binary(16)",
        foreignKey = ForeignKey(name = "fk_order_line_item_to_orders"),
    )
    val orderLineItems: List<OrderLineItemV2>,
) {
    companion object {
        fun from(orderLineItems: List<OrderLineItemV2>): OrderLineItems {
            return OrderLineItems(orderLineItems = orderLineItems)
        }
    }
}
